eth/filters/filter_system_test.go

```diff
@@ -601,6 +606,60 @@ func TestPendingLogsSubscription(t *testing.T) {
 	}
 }
 
+// TestPendingTxFilterDeadlock tests if the event loop hangs when pending
+// txes arrive at the same time that one of multiple filters is timing out.
+func TestPendingTxFilterDeadlock(t *testing.T) {
+	t.Parallel()
+	timeout := 100 * time.Millisecond
+
+	var (
+		db      = rawdb.NewMemoryDatabase()
+		backend = &testBackend{db: db}
+		api     = NewPublicFilterAPI(backend, false, timeout)
+		done    = make(chan struct{})
+	)
+
+	go func() {
+		// Bombard feed with txes until signal was received to stop
+		i := uint64(0)
+		for {
+			select {
+			case <-done:
+				return
+			default:
+			}
+
+			tx := types.NewTransaction(i, common.HexToAddress("0xb794f5ea0ba39494ce83a213fffba74279579268"), new(big.Int), 0, new(big.Int), nil)
+			backend.txFeed.Send(core.NewTxsEvent{Txs: []*types.Transaction{tx}})
+			i++
+		}
+	}()
+
+	// Create a bunch of filters that will
+	// timeout either in 100ms or 200ms
+	for i := 0; i < 20; i++ {
+		fid := api.NewPendingTransactionFilter()
+		// Wait for at least one tx to arrive in filter
+		for {
+			if _, err := api.GetFilterChanges(fid); err == nil {
+				break
+			}
+			runtime.Gosched()
+		}
+	}
+
+	// Wait until filters have timed out
+	time.Sleep(250 * time.Millisecond)
+
```

Comment:
We should also check that all created filters have actually timed out.

Reply:
Ah on second thought that comment is misleading. At least one filter will time out but we won't know for sure exactly how many will, because during the unsubscription of one of them the routine will hang.

Should I check to see if at least one has been uninstalled (by checking `api.filters` length?)

Reply:
If the deadlock does not happen, all filters should be gone after this sleep. We should check this because it's an invariant here. You can check this by running `GetFilterChanges` for all filters (you'll need to collect their IDs into a slice for that). The `GetFilterChanges` call should return an error for all of them.

Reply:
Makes sense. Added the check in the select clause's happy case.

Reply:
Why though? We can always check this, right? Worst case it will produce another error message from the test. It's OK to raise more than one error from a test.

Reply:
You're right I don't know why my brain couldn't process a test failing twice :)

I just tried this but there's another problem. when a deadlock happens `GetFilterChanges` also hangs. One option I can think of is to go unsafe and directly read `len(api.filters)` without taking the lock. Wdyt?